File: aes/aes.py
from .aes_helper import (expand_key, byte_to_state, add_round_key, create_round_key,
                        sub_bytes, shift_rows, mix_columns, state_to_byte,
                        inv_shift_rows, inv_sub_bytes, inv_mix_columns)


class AES:
    """Implementation of the AES encryption algorithm."""
    def __init__(self, key_size: int = 16):
        if key_size not in (16, 24, 32):
            raise ValueError("key_size must be 16, 24, or 32 (bytes).")
        self.key_size = key_size # in bytes
        self._rounds = {16: 10, 24: 12, 32: 14}[key_size]
        # self.modified_aes_512 = False  # to indicate if AES-512 is used


    def encrypt(self, input_bytes: bytearray, key: bytearray) -> bytearray:
        """Encrypt a 16-byte block with AES using the given key."""
        expanded_key = expand_key(key, self.key_size, 16 * (self._rounds + 1))
        state = byte_to_state(input_bytes)

        # Initial Round
        state = add_round_key(state, create_round_key(expanded_key, 0))

        # Main Rounds
        for round_idx in range(1, self._rounds):
            state = sub_bytes(state)
            state = shift_rows(state)
            state = mix_columns(state)
            state = add_round_key(state, create_round_key(expanded_key, round_idx))

        # Final Round (no MixColumns)
        state = sub_bytes(state)
        state = shift_rows(state)
        state = add_round_key(state, create_round_key(expanded_key, self._rounds))

        return state_to_byte(state)


    def decrypt(self, input_bytes: bytearray, key: bytearray) -> bytearray:
        """Decrypt a 16-byte block with AES using the given key."""
        expanded_key = expand_key(key, self.key_size, 16 * (self._rounds + 1))
        state = byte_to_state(input_bytes)

        # Initial Round
        state = add_round_key(state, create_round_key(expanded_key, self._rounds))

        # Main Rounds
        for round_idx in range(self._rounds - 1, 0, -1):
            state = inv_shift_rows(state)
            state = inv_sub_bytes(state)
            state = add_round_key(state, create_round_key(expanded_key, round_idx))
            state = inv_mix_columns(state)

        # Final Round (no InvMixColumns)
        state = inv_shift_rows(state)
        state = inv_sub_bytes(state)
        state = add_round_key(state, create_round_key(expanded_key, 0))

        return state_to_byte(state)

File: src_mat/aes/AES.m
classdef AES
    properties (Access = private)
        key_size
        rounds
        expanded_key
    end

    methods
        function obj = AES(key)
            key = uint8(key(:).');  % row vector
            obj.key_size = numel(key);

            if ~ismember(obj.key_size, [16 24 32])
                error('key_size must be 16, 24, or 32 bytes.');
            end

            switch obj.key_size
                case 16
                    obj.rounds = 10;
                case 24
                    obj.rounds = 12;
                case 32
                    obj.rounds = 14;
            end

            expanded_key_size = 16 * (obj.rounds + 1);
            obj.expanded_key = aes_helper.expand_key( ...
                key, obj.key_size, expanded_key_size);
        end

        function out = encrypt(obj, input_bytes)

            input_bytes = uint8(input_bytes(:).');
            if numel(input_bytes) ~= 16
                error('Input block must be 16 bytes.');
            end

            state = aes_helper.byte_to_state(input_bytes);

            % Initial round
            state = aes_helper.add_round_key( ...
                state, aes_helper.create_round_key(obj.expanded_key, 0));

            % Main rounds
            for round_idx = 1:(obj.rounds - 1)
                state = aes_helper.sub_bytes(state);
                state = aes_helper.shift_rows(state);
                state = aes_helper.mix_columns(state);
                state = aes_helper.add_round_key( ...
                    state, aes_helper.create_round_key(obj.expanded_key, round_idx));
            end

            % Last round
            state = aes_helper.sub_bytes(state);
            state = aes_helper.shift_rows(state);
            state = aes_helper.add_round_key( ...
                state, aes_helper.create_round_key(obj.expanded_key, obj.rounds));

            out = aes_helper.state_to_byte(state);
        end

        function out = decrypt(obj, input_bytes)

            input_bytes = uint8(input_bytes(:).');
            if numel(input_bytes) ~= 16
                error('Input block must be 16 bytes.');
            end

            state = aes_helper.byte_to_state(input_bytes);

            % Initial round
            state = aes_helper.add_round_key( ...
                state, aes_helper.create_round_key(obj.expanded_key, obj.rounds));

            % First round
            for round_idx = (obj.rounds - 1):-1:1
                state = aes_helper.inv_shift_rows(state);
                state = aes_helper.inv_sub_bytes(state);
                state = aes_helper.add_round_key( ...
                    state, aes_helper.create_round_key(obj.expanded_key, round_idx));
                state = aes_helper.inv_mix_columns(state);
            end

            % Last round
            state = aes_helper.inv_shift_rows(state);
            state = aes_helper.inv_sub_bytes(state);
            state = aes_helper.add_round_key( ...
                state, aes_helper.create_round_key(obj.expanded_key, 0));

            out = aes_helper.state_to_byte(state);
        end
    end
end
